cmd/abapEnvironmentAssembly.go

```diff
@@ -0,0 +1,598 @@
+	piperhttp "github.com/SAP/jenkins-library/pkg/http"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/telemetry"
+	"github.com/pkg/errors"
+)
+
+func abapEnvironmentAssembly(config abapEnvironmentAssemblyOptions, telemetryData *telemetry.CustomData, commonPipelineEnvironment *abapEnvironmentAssemblyCommonPipelineEnvironment) {
+	// for command execution use Command
+	c := command.Command{}
+	// reroute command output to logging framework
+	c.Stdout(log.Writer())
+	c.Stderr(log.Writer())
+
+	// for http calls import  piperhttp "github.com/SAP/jenkins-library/pkg/http"
+	// and use a  &piperhttp.Client{} in a custom system
+	// Example: step checkmarxExecuteScan.go
+
+	// error situations should stop execution through log.Entry().Fatal() call which leads to an os.Exit(1) in the end
+	err := runAbapEnvironmentAssembly(&config, telemetryData, &c, commonPipelineEnvironment)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+
+func runAbapEnvironmentAssembly(config *abapEnvironmentAssemblyOptions, telemetryData *telemetry.CustomData, command command.ExecRunner, commonPipelineEnvironment *abapEnvironmentAssemblyCommonPipelineEnvironment) error {
+	return runAssembly(config)
+}
+
+// *******************************************************************************************************************************
+// ********************************************************** Step logic *********************************************************
+// *******************************************************************************************************************************
+func runAssembly(config *abapEnvironmentAssemblyOptions) error {
+	conn := new(connector)
+	err := conn.init(*config, &piperhttp.Client{})
+	if err != nil {
+		return err
+	}
+	assemblyBuild := build{
+		connector: *conn,
+	}
+
+	valuesInput := values{
+		Values: []value{
+			{
+				ValueID: "SWC",
+				Value:   config.SWC,
+			},
+			{
+				ValueID: "CVERS",
+				Value:   config.SWC + "." + config.SWCRelease + "." + config.SpsLevel,
+			},
+			{
+				ValueID: "NAMESPACE",
+				Value:   config.Namespace,
+			},
+			{
+				ValueID: "PACKAGE_NAME_" + config.PackageType,
+				Value:   config.PackageName,
+			},
+		},
+	}
+	if config.PreviousDeliveryCommit != "" {
+		valuesInput.Values = append(valuesInput.Values,
+			value{ValueID: "PREVIOUS_DELIVERY_COMMIT",
+				Value: config.PreviousDeliveryCommit})
+	}
+	phase := "BUILD_" + config.PackageType
+	err = assemblyBuild.startPollLog(phase, valuesInput, time.Duration(config.MaxRuntimeInMinutes), 60)
+	if err != nil {
+		return err
+	}
+	resultName := "SAR_XML"
+	resultSARXML, err := assemblyBuild.getResult(resultName)
+	if err != nil {
+		return err
+	}
+	envPath := filepath.Join(GeneralConfig.EnvRootPath, "commonPipelineEnvironment")
+	downloadPath := filepath.Join(envPath, path.Base(resultName))
+	err = resultSARXML.download(downloadPath)
+	return err
```

Comment:
I'm not that familiar with the commonPipelineEnvironment. But, if possible, use "commonPipelineEnvironment *abapEnvironmentAssemblyCommonPipelineEnvironment" (Signature of runAbapEnvironmentAssembly) to write to the commonPipelineEnvironment.

Reply:
I rewrote this. Now I am still downloading the SAR_XML file to the commonPipelineEnvironment folder, I just needed a place to download it, where it is ensured that the next step can upload it again. However what I add to the real commonPipelineEnvironment is the path to this file, as there can be several SAR_XML files (one for each repository/CV in the addon.yml file)